src/ModelContextProtocol/Protocol/Transport/StdioServerTransport.cs

```diff
@@ -73,4 +73,32 @@ private static string GetServerName(McpServerOptions serverOptions)
 
         return serverOptions.ServerInfo?.Name ?? McpServer.DefaultImplementation.Name;
     }
+
+    // Neither WindowsConsoleStream nor UnixConsoleStream respect CancellationTokens or cancel any I/O on Dispose.
+    // WindowsConsoleStream will return an EOS on Ctrl-C, but that is not the only reason the shutdownToken may fire.
+    private sealed class CancellableStdinStream(Stream stdinStream) : Stream
+    {
+        public override bool CanRead => true;
+        public override bool CanSeek => false;
+        public override bool CanWrite => false;
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => stdinStream.ReadAsync(buffer, offset, count, cancellationToken).WaitAsync(cancellationToken);
```

Comment:
These will not be recoverable, i.e. the actual read is not being canceled, just the wait, so if there's ever more data on the stream, the pending read will eat it, and any subsequent reads will miss that data.

We will need to ensure that cancellation tears down the server and this stream isn't used again.

Reply:
Agreed. That's why I limited this change to StdioServerTransport which typically is only disposed when tearing down the process and didn't make any changes to the base StreamServerTransport. Obviously, it'd be better if the stream returned by Console.OpenStandardInput() allowed cancelling the read operation propery in any manner, either with the token or by calling Dispose mid-operation, but neither of those things currently work.